### PR TITLE
Tock 2.0: update ADC driver to 2.0 API

### PR DESCRIPTION
**Pull request by alexandruradovici**
### Pull Request Overview

This pull request updates the ADC driver to Tock 2.0 API.

### Testing Strategy

I tested the simple ADC using a Microbit v2.
High Speed ADC has not yet been tested.

### TODO or Help Wanted

As I have no board that supports the high speed ADC, @hudson-ayers offered to help and test it. A PR with the new API is available (tock/libtock-c#149).

As the high speed part has some tricky code parts, this PR needs a careful review, mostly for the buffer filling part.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/src/adc.rs [hunk 1]
```diff
@@ -963,70 +962,79 @@ impl<A: hil::adc::Adc + hil::adc::AdcHighSpeed> hil::adc::HighSpeedClient for Ad
                                 //            or sample length
                                 // We then split each sample into its two bytes and copy
                                 // them to the app buffer
-                                for (chunk, &sample) in app_buf
-                                    .chunks_mut(2)
+
+                                let mut position = 0;
+                                for &sample in adc_buf
+                                    .iter()
```
[the remark this answers is not in the document]
Comment (hudson-ayers) [reply capsules/src/adc.rs:968]: also, the new logic leads to an out-of-bounds panic when I run the adc app on Imix, but the old logic seems to work correctly

**Review by hudson-ayers — CHANGES_REQUESTED** [review]
Thanks so much for taking this one on!

I tested on Imix and got a kernel panic, but if I change a small amount of logic back to how it was before it seems to work correctly. I also have a few other comments

File: capsules/src/adc.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs:968]: I changed it as it was using chucks which are not available anymore. I will double check it.

File: capsules/src/adc.rs [hunk 2]
```diff
@@ -1248,50 +1269,86 @@ impl<A: hil::adc::Adc + hil::adc::AdcHighSpeed> LegacyDriver for AdcDedicated<'_
...
-            5 => self.stop_sampling(),
+            5 => match self.stop_sampling() {
+                ReturnCode::SUCCESS => CommandResult::success(),
+                ReturnCode::SuccessWithValue { value } => CommandResult::success_u32(value as u32),
+                e => CommandResult::failure(if let Ok(err) = ErrorCode::try_from(e) {
+                    err
+                } else {
+                    panic!("ADC: invalid return code")
+                }),
+            },
```
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs:1331]: Got it

File: capsules/src/adc.rs [hunk 3]
```diff
@@ -697,14 +701,12 @@ impl<A: hil::adc::Adc + hil::adc::AdcHighSpeed> hil::adc::Client for AdcDedicate
...
                     .enter(*id, |app, _| {
-                        app.callback.map(|callback| {
-                            calledback = true;
-                            callback.schedule(
-                                AdcMode::SingleSample as usize,
-                                self.channel.get(),
-                                sample as usize,
-                            );
-                        });
+                        calledback = true;
```
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs:704]: I don't see any other way of doing this.

File: capsules/src/adc.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs]: I think I fixed the overflow, let me know if it works.

As a not here, I'm not sure what happens if the user application changes the buffer in between two sample buffers. The `len()` will change and might be lower than the driver expects it to be.

File: capsules/src/adc.rs [hunk 4]
```diff
@@ -963,70 +962,79 @@ impl<A: hil::adc::Adc + hil::adc::AdcHighSpeed> hil::adc::HighSpeedClient for Ad
...
+                                    self.replace_buffer(buf);
+                                });
+                            } else {
+                                // if the mode is ContinuousBuffer, we've just
+                                // switched app buffers. Reset our offset to zero
+                                app.app_buf_offset.set(0);
                             }
-                        });
+                        }
+                        // });
```
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs]: Done

File: capsules/src/adc.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs]: It turns out that none of the functions returns `SuccessWithValue`, so I just deleted the match arm.

File: capsules/src/adc.rs [hunk 5]
```diff
@@ -338,8 +342,8 @@ impl<'a, A: hil::adc::Adc + hil::adc::AdcHighSpeed> AdcDedicated<'a, A> {
         let exists = self.appid.map_or(false, |id| {
             self.apps
                 .enter(*id, |state, _| {
-                    app_buf_length = state.app_buf1.as_mut().map_or(0, |buf| buf.len());
-                    state.app_buf1.is_some()
+                    app_buf_length = state.app_buf1.len();
+                    state.app_buf1.len() > 0
```
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs]: It is still strange why it panics when getting the length via the `len()` function. Inside the SharedSlice, it does the same map.

File: capsules/src/adc.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hudson-ayers) [reply capsules/src/adc.rs]: why are chunks not available anymore?

[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs]: I might be missing something, `chunks` is implemented for `AppSlice` (https://github.com/tock/tock/blob/195e0061ba1a186dc3a38e388ee2a6d0372dacea/kernel/src/mem.rs#L417), but  not for `ReadWriteAppSlice`.

[the remark this answers is not in the document]
Comment (alexandruradovici) [reply capsules/src/adc.rs]: Reverted.

File: capsules/src/adc.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hudson-ayers) [reply capsules/src/adc.rs:704]: It doesn't look like userspace relied on this, so I think this is okay as-is.

**Review by hudson-ayers — APPROVED** [review]

**Pull request merged**